Posts/2021/02/ScriptingGuy.0002.GetDateOfYesterday.md

```diff
@@ -0,0 +1,185 @@
+
+PS C:\> # Or more simply
+PS C:\> $Yesterday = (Get-Date).AddDays(-1)
+PS C:\> $Yesterday
+19 February 2021 12:13:52
+
+PS C:> # Get tomorrow's date
+PS C:> $Tomorrow = (Get-Date).AddDays(1)
+PS C:> $Tomorrow
+21 February 2021 12:13:54
+```
+It is worth noting that a `System.DateTime` object is immutable.
+This means you can not change propery values after you create the object.
+If you use any of the `Add` methods, .NET returns a new object with updated property values. 
+
+## Using Yesterday's Date
+
+There are a variety use cases for getting a date in the past (or the future), including 
+**Identifying files that are older/younger than a day/month/etc ago
+**Determining which AD Users have not logged on in the last week
+**Creating a file name for a file representing last weeks information.
+
+Here are some examples:
+
+```powershell-console
+PS C:> # Finding files newer than yesterday
+PS C:> $Yesterday = (Get-Date).AddDays(-1)
+PS C:> Get-ChildItem | Where-Object LastAccessTime -gt $Yesterday
+
+    Directory: C:\
+    
+Mode                 LastWriteTime         Length Name
+----                 -------------         ------ ----
+-a---          20/02/2021    14:20          11041 GratefulDead Show List.txt
+
+PS C:> # Getting users who have logged on in the past day
+PS C:> Get-ADUser -Filter * -Property LastLogonDate | Where LastlogonDate -gt $Yesterday
+
+DistinguishedName : CN=Administrator,CN=Users,DC=cookham,DC=net
+Enabled           : True
+GivenName         : Jerry
+LastLogonDate     : 20/02/2021 04:20:42
+Name              : Jerry Garcia
+ObjectClass       : user
+ObjectGUID        : ae31ca0d-3f01-4eb4-8593-b1d79c71f912
+SamAccountName    : JerryG
+SID               : S-1-5-21-2550804810-443649076-1856842782-500
+Surname           : Garcia
+
+# Creating a file with yesterday's date
+PS C:\> # Creating a file with today's date
+PS C:\> $Yesterday     = (Get-Date).AddDays(-1).ToString() -replace '/','-'
+PS C:\> $YesterDayDate = ($Yesterday -split ' ')[0]
+PS C:\> $YesterdayFN   = "Results for $YesterdayDate.Txt"
+PS C:\> 
+PS C:\> New-Item -Path C:\Results -Name  $YesterdayFN -ItemType File
+
+Directory: C:\Results
+
+Mode                 LastWriteTime         Length Name
+----                 -------------         ------ ----
+-a---          20/02/2021    12:56              0 Results for 19-02-2021.Txt
+```
+
+In that last example, you need to do a bit of manipulation of the date/time returned by `Get-Date` in order to get a filename that Windows accepts.
+This manipulation is needed because `Get-Date` returns a string that contains the "/" character which confuses `New-Item`.
+You use the `-Replace` operator to replace the "/" character with a "-".
+Additionally, after performing the replacement, you end up with an (unneeded) time value.
+You can use the `-Split` operator to pull out just the date, which is what you want for the file name.
+Once you do get the date, you can create you can create a file name for the file,
```

Comment:
Will this work on all cultures? I would recommend using an example like this `(Get-Date).ToString('dd-MM-yyyy')`. This will remove the extra steps and confusion on why they need to change it from `/` to '-'. For further reading you could link out to the `ToString()` method documentation for further format options or this could also be a separate blog.

Reply:
Get-Date does a pretty good job. I've made a comment in the article to this effect.

Reply:
If you're going to leave it as is. You'll need to change "This manipulation is needed because `Get-Date` returns a string that contains the "/" character which confuses `New-Item`." to actually say why `/` is not allowed. The word "confused" is not the correct term and is confusing its self. It doesn't work because `/` is a path separator character.